### PR TITLE
virtual device : modify virtual device framework by deleting dependency

### PR DESCRIPTION
**Pull request by joohyun0115**
Modify virtual device framework by deleting dependency

### CODE REVIEW

**Comment by kesl** [comment]
CI : Test FAILed.
Refer to this link for build results: http://keslproject.iptime.org:8082/job/k-hypervisor_pull_request/124/

**Comment by joohyun0115** [comment]
retest this please

**Comment by kesl** [comment]
CI : Test PASSed.
Refer to this link for build results: http://keslproject.iptime.org:8082/job/k-hypervisor_pull_request/125/

File: hypervisor/include/vdev.h [hunk 1]
```diff
@@ -11,15 +11,31 @@ enum vdev_access_size {
...
 };
 
-/*
- * @write   0: read, 1: write
- * @offset  offset in bytes from base address
- * @pvalue  address to input value (if write) or output value (if read)
- * return: HVMM_STATUS_SUCCESS if successful, failed otherwise
- */
+enum vdev_level {
+    VDEV_LEVEL_LOW = 0,
```
Comment (Wonseok) [inline hypervisor/include/vdev.h:15]: You don't need to use '= 0'.

File: hypervisor/include/vdev.h [hunk 2]
```diff
@@ -11,15 +11,31 @@ enum vdev_access_size {
     VDEV_ACCESS_RESERVED = 3
 };
 
-/*
- * @write   0: read, 1: write
- * @offset  offset in bytes from base address
- * @pvalue  address to input value (if write) or output value (if read)
- * return: HVMM_STATUS_SUCCESS if successful, failed otherwise
- */
+enum vdev_level {
```
Comment (Wonseok) [inline hypervisor/include/vdev.h:14]: Please explain the level of vdev.

File: hypervisor/include/vdev.h [hunk 3]
```diff
@@ -27,11 +43,95 @@ struct vdev_info {
...
+    hvmm_status_t (*disable)(void);
+
+    /** Currently unused function */
+    hvmm_status_t (*request_irq)(vdev_irq_callback_t, void *user);
+
+    /** Currently unused function */
+    hvmm_status_t (*notify_irq)(uint32_t);
+
+    /** Post for remain on the job */
+    hvmm_status_t (*post)(struct arch_vdev_info *, struct arch_regs *);
```
Comment (Wonseok) [inline hypervisor/include/vdev.h:90]: Why do we need to post for remain jobs? Please explain what post do.

File: hypervisor/vdev.c [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+ * \retval 0 if virtual device is registed correctly
+ * \retval -3 This is an internal error
+ */
+hvmm_status_t vdev_register(int level, struct vdev_module *module)
+{
+    int i;
+    hvmm_status_t result = HVMM_STATUS_BUSY;
+
+    for (i = 0; i < MAX_VDEV; i++) {
+        if (!_vdev_module[level][i]) {
```
Comment (Wonseok) [inline hypervisor/vdev.c:26]: Please do not use !, sometimes it makes us confused. You might want to use such as "if (_vdev_module[level][i] != 0 or something)".

File: hypervisor/vdev.c [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+}
+
+/**
+ * \brief Lookup the virtual deivce address, using the archtecture specific
+ * information \a info and current archtecture specific register \a regs.
+ *
+ * \retval virtual device number
+ * \retval -1 This is an internal error.
+ */
+int32_t vdev_search(int level, struct arch_vdev_info *info,
```
Comment (Wonseok) [inline hypervisor/vdev.c:48]: s/search/find

File: hypervisor/vdev.c [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+        struct arch_regs *regs)
+{
+    int32_t i;
+    int32_t vdev_num = -1;
```
Comment (Wonseok) [inline hypervisor/vdev.c:52]: You might want to use named constant.

File: hypervisor/vdev.c [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+int32_t vdev_read(int level, int num, struct arch_vdev_info *info,
+            struct arch_regs *regs)
+{
+    int32_t size = 0;
+    struct vdev_module *vdev = _vdev_module[level][num];
+
+    if (!vdev) {
+        printh("vdev : Could not get module, level : %d, i : %d\n",
+                level, num);
+        return -1;
```
Comment (Wonseok) [inline hypervisor/vdev.c:87]: Please use named constant instead of real value.

File: hypervisor/vdev.c [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+    }
+
+    return vdev_num;
+}
+
+int32_t vdev_read(int level, int num, struct arch_vdev_info *info,
+            struct arch_regs *regs)
+{
+    int32_t size = 0;
+    struct vdev_module *vdev = _vdev_module[level][num];
```
Comment (Wonseok) [inline hypervisor/vdev.c:82]: What does num means?

File: hypervisor/vdev.c [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+        return -1;
+    }
+
+    if (vdev->ops->write)
+        size = vdev->ops->write(info, regs);
+
+    return size;
+}
+
+hvmm_status_t vdev_post(int level, int num, struct arch_vdev_info *info,
```
Comment (Wonseok) [inline hypervisor/vdev.c:126]: arch_vdev is not good naming but I don't have an idea better than this.
s/arch_vdev_info/dabort_info/
Anyway info is not good word for naming. If you have an idea, please change it.

**Comment by Wonseok** [comment]
Please use named constant instead of real value because it makes us confused what function did.
You might want to use "#define ERROR  -1 and return ERROR" instead of "return -1".
Naming is not good currently, please make it clarify.

File: hypervisor/include/vdev.h (hunk 2 above)
Comment (joohyun0115) [inline hypervisor/include/vdev.h:14]: That will be performance improvement. when searching for a virtual device. 
ARM :  
level one : co-processor & sensitive instruction.
level two : hyper-call
level three : device area

File: hypervisor/include/vdev.h (hunk 3 above)
Comment (joohyun0115) [inline hypervisor/include/vdev.h:90]: Some device need for remain job like a below.

[gicd]
static int32_t vdev_gicd_post(struct arch_vdev_info *info,
                        struct arch_regs *regs)
{
    uint8_t isize = 4;

```
if (regs->cpsr & 0x20) /* Thumb */
    isize = 2;

regs->pc += isize;

return 0;
```

}

File: hypervisor/vdev.c (hunk 4 above)
Comment (joohyun0115) [inline hypervisor/vdev.c:26]: Open source area usually using the !xxxx more than ...!= 0.. such as linux kenrel & xen..

File: hypervisor/vdev.c (hunk 5 above)
Comment (joohyun0115) [inline hypervisor/vdev.c:48]: Ok

File: hypervisor/vdev.c (hunk 8 above)
Comment (joohyun0115) [inline hypervisor/vdev.c:82]: Virtual device position in array.

File: hypervisor/vdev.c (hunk 7 above)
Comment (joohyun0115) [inline hypervisor/vdev.c:87]: Do you want to constant value? like a hvmm_status_t?
But this value indicate a size. 
We do not have constant value for general error.  ㅠ.ㅠ

File: hypervisor/vdev/vdev_hvc_stay.c [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+
+static int32_t vdev_hvc_stay_write(struct arch_vdev_info *info,
+                        struct arch_regs *regs)
+{
+    printh("[hyp] _hyp_hvc_service:stay\n\r");
+    context_dump_regs(regs);
+
+    return 0;
+}
+static int32_t vdev_hvc_stay_check(struct arch_vdev_info *info,
```
Comment (Wonseok) [inline hypervisor/vdev/vdev_hvc_stay.c:18]: Add empty line a function above.

**Pull request closed without merging**